### PR TITLE
Chef 13 Compatibility Fix

### PR DESCRIPTION
**Pull request by JerryPeng0112**

### CODE REVIEW

File: Berksfile [hunk 1]
```diff
@@ -1,6 +1,7 @@
 source 'https://supermarket.chef.io'
 
 # cookbook 'percona', '>= 0.15.0'
+cookbook 'base', git: 'git@github.com:osuosl-cookbooks/base'
 cookbook 'percona', github: 'phlipper/chef-percona'
```
Comment (ramereth) [inline Berksfile]: Let's no longer rely on the github repo and remove this and the reference above.

File: metadata.rb [hunk 2]
```diff
@@ -17,6 +19,7 @@
 depends          'percona'
```
Comment (ramereth) [inline metadata.rb]: Please add a lock with ``~> 0.16.1`` which matches what we have in production too.

File: recipes/server.rb [hunk 3]
```diff
@@ -46,7 +46,10 @@
 node.default['percona']['server']['innodb_buffer_pool_size'] =
   "#{Integer(mem * 0.75)}M"
 
-include_recipe 'sysctl::default'
+include_recipe 'yum-epel'
+package 'libev'
```
Comment (ramereth) [inline recipes/server.rb]: Why is this packaged needed?

File: recipes/server.rb [hunk 4]
```diff
@@ -72,6 +76,8 @@
   mode '0755'
 end
 
+package 'cronie'
```
Comment (ramereth) [inline recipes/server.rb]: Why is this needed?

File: recipes/xtrabackuprb.rb [hunk 5]
```diff
@@ -1,5 +1,12 @@
 include_recipe 'git'
 include_recipe 'percona::package_repo'
+include_recipe 'yum-epel'
+
+package 'rubygems' do
+end.run_action(:install)
```
Comment (ramereth) [inline recipes/xtrabackuprb.rb]: This isn't needed since we're using the chef rubygems.

File: metadata.rb [hunk 6]
```diff
@@ -17,6 +19,7 @@
...
 depends          'sysctl'
 depends          'yum'
+depends          'yum-epel'
```
Comment (ramereth) [inline metadata.rb]: What required added EPEL?

File: recipes/xtrabackuprb.rb [hunk 7]
```diff
@@ -15,10 +22,9 @@
 
 chef_gem 'xtrabackup-rb' do
   source "/usr/local/src/xtrabackup-rb/xtrabackup-rb-#{version}.gem"
+  compile_time false
```
Comment (ramereth) [inline recipes/xtrabackuprb.rb]: Why is this now set to false?

File: files/default/mysql-accounting [hunk 8]
```diff
@@ -1,7 +1,7 @@
 #!/bin/bash
 # Output current disk usage for each database
 DATE=$(date +%Y%m%d)
-ACCOUNTING_DIR="/var/lib/accounting/mysql"
+ACCOUNTING_DIR="/var/lib/mysql-files"
```
Comment (ramereth) [inline files/default/mysql-accounting]: This path should **not** be changed. Why was it changed?

File: recipes/server.rb [hunk 9]
```diff
@@ -63,7 +66,8 @@
   only_if { platform_family?('rhel') }
 end
 
-sysctl_param 'vm.swappiness' do
+sysctl_param 'set vm.swappiness to 0' do
+  key 'vm.swappiness'
```
Comment (ramereth) [inline recipes/server.rb]: Just have the resource name be ``vm.swappiness``, then setting ``key`` is no longer required.

File: Berksfile (hunk 1 above)
Comment (JerryPeng0112) [reply Berksfile]: @ramereth Do you mean line 3-5?
For line 4, if it is deleted, chef uses base cookbook from the supermarket.

File: metadata.rb (hunk 6 above)
Comment (JerryPeng0112) [reply metadata.rb]: For server recipe, percona-xtrabackup requires libev package, so I included yum-epel to install libev

File: recipes/server.rb (hunk 9 above)
Comment (JerryPeng0112) [reply recipes/server.rb]: @ramereth It has cloning conflict with base::sysctl since base::sysctl also has resource that set vm.swappiness. I've reverted it.

File: Berksfile (hunk 1 above)
Comment (ramereth) [reply Berksfile]: I mean the line which contains ``percona``

File: files/default/mysql-accounting (hunk 8 above)
Comment (JerryPeng0112) [reply files/default/mysql-accounting]: @ramereth I set to this directory because secure_file_priv is set to that directory. If this is set, mysql can only write to that directory. Should I disable secure_file_priv, or maybe change the secure_file_priv to "/var/lib/accounting/mysql"?

File: recipes/xtrabackuprb.rb (hunk 7 above)
Comment (JerryPeng0112) [reply recipes/xtrabackuprb.rb]: Set back to true in latest commit. Chef 13 requires compile_time attribute to be specified

File: recipes/xtrabackuprb.rb (hunk 5 above)
Comment (JerryPeng0112) [reply recipes/xtrabackuprb.rb]: This is an error I made because I accidentally make `chef_gem 'xtrabackup-rb'` to compile_time false, went down the rabbit hole and installed rubygems. It is now corrected.

Comment (JerryPeng0112) [reply recipes/xtrabackuprb.rb]: However, we will need 'libev' package for 'percona-xtrabackup', so still need to include_recipe 'yum-epel' for installing 'libev'

File: files/default/mysql-accounting (hunk 8 above)
Comment (ramereth) [reply files/default/mysql-accounting]: Are you referring to selinux settings? AFAIK this script runs as root so it shouldn't matter.

File: Berksfile (hunk 1 above)
Comment (ramereth) [reply Berksfile]: @JerryPeng0112 you still need to remove this line.

File: recipes/xtrabackuprb.rb [hunk 10]
```diff
@@ -1,24 +1,30 @@
 include_recipe 'git'
 include_recipe 'percona::package_repo'
 
+if platform_family?('rhel') && node['platform_version'].to_i == 6
+  include_recipe 'yum-epel'
+  package 'libev'
```
Comment (ramereth) [inline recipes/xtrabackuprb.rb]: There's no need to install the ``libev`` package since that get's pulled in via the ``percona-xtrabackup`` package as a dependency. I recommend you just change this whole block to be:
``` ruby
include_recipe 'yum-epel' if node['platform_version'].to_i == 6
￼```

File: spec/mon_spec.rb [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+      end
+      it do
+        expect(chef_run).to include_recipe('percona::monitoring')
+      end
+      it do
+        expect(chef_run).to include_recipe('osl-nrpe')
+      end
+      it do
+        expect(chef_run).to include_recipe('osl-munin::client')
+      end
```
Comment (ramereth) [inline spec/mon_spec.rb]: Use an array for testing all of these recipes being included.

File: recipes/xtrabackuprb.rb [hunk 12]
```diff
@@ -1,24 +1,30 @@
...
+end
+
+package 'percona-xtrabackup'
```
Comment (ramereth) [inline recipes/xtrabackuprb.rb]: This installation fails since we now don't include postfix on the vms. I recommend you include the ``osl-postfix`` cookbook above to work around this.

File: spec/server_spec.rb [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+      end
+      it do
+        expect(chef_run).to include_recipe('percona::toolkit')
+      end
+      it do
+        expect(chef_run).to include_recipe('percona::backup')
+      end
+      it do
+        expect(chef_run).to include_recipe('firewall::mysql')
+      end
```
Comment (ramereth) [inline spec/server_spec.rb]: Use a loop for these recipes.

File: files/default/mysql-accounting (hunk 8 above)
Comment (JerryPeng0112) [reply files/default/mysql-accounting]: @ramereth Not sure, it would not converge, however. I got my solution [here](https://stackoverflow.com/questions/34102562/mysql-error-1290-hy000-secure-file-priv-option)

File: recipes/xtrabackuprb.rb [hunk 14]
```diff
@@ -1,24 +1,25 @@
 include_recipe 'git'
 include_recipe 'percona::package_repo'
 
+package 'percona-xtrabackup'
```
Comment (ramereth) [inline recipes/xtrabackuprb.rb]: Above here you need to add ``include_recipe 'yum-epel' if node['platform_version'].to_i == 6``

**Comment by ramereth** [comment]
@JerryPeng0112 I think you can fix the mysql accounting file with the following:

``` diff
diff --git a/files/default/mysql-accounting b/files/default/mysql-accounting
index 1664e36..376740b 100644
--- a/files/default/mysql-accounting
+++ b/files/default/mysql-accounting
@@ -1,13 +1,15 @@
-#!/bin/bash
+#!/bin/bash -eu
 # Output current disk usage for each database
 DATE=$(date +%Y%m%d)
 ACCOUNTING_DIR="/var/lib/accounting/mysql"
-mkdir -p $ACCOUNTING_DIR
-chown mysql:mysql $ACCOUNTING_DIR
+if [ -e "/var/lib/mysql-files/${DATE}.csv" ] ; then
+  rm -f /var/lib/mysql-files/${DATE}.csv
+fi
 mysql --defaults-extra-file=/root/.my.cnf -e "SELECT table_schema 'DB Name',
   round(sum( data_length + index_length ) / 1024 / 1024, 1) 'Data Base Size in MB'
   FROM information_schema.TABLES
   GROUP BY table_schema
-  INTO OUTFILE '${ACCOUNTING_DIR}/${DATE}.csv'
+  INTO OUTFILE '/var/lib/mysql-files/${DATE}.csv'
   FIELDS TERMINATED BY ',' ENCLOSED BY '\"' LINES TERMINATED BY '\n';"
-gzip ${ACCOUNTING_DIR}/${DATE}.csv
+mv -f /var/lib/mysql-files/${DATE}.csv ${ACCOUNTING_DIR}/${DATE}.csv
+gzip -f ${ACCOUNTING_DIR}/${DATE}.csv
diff --git a/recipes/server.rb b/recipes/server.rb
index 6396f40..305c8e7 100644
--- a/recipes/server.rb
+++ b/recipes/server.rb
@@ -67,6 +67,16 @@ sysctl_param 'vm.swappiness' do
   value 0
 end
 
+directory '/var/lib/mysql-files' do
+  owner 'mysql'
+  group 'mysql'
+end
+
+directory '/var/lib/accounting/mysql' do
+  recursive true
+  mode '0700'
+end
+
 cookbook_file '/usr/local/libexec/mysql-accounting' do
   source 'mysql-accounting'
   mode '0755'
diff --git a/test/integration/mon/serverspec/mon_spec.rb b/test/integration/mon/serverspec/mon_spec.rb
index 035bfc8..2528034 100644
--- a/test/integration/mon/serverspec/mon_spec.rb
+++ b/test/integration/mon/serverspec/mon_spec.rb
@@ -56,3 +56,7 @@ end
     its(:exit_status) { should eq 0 }
   end
 end
+
+describe command('/usr/local/libexec/mysql-accounting') do
+  its(:exit_status) { should eq 0 }
+end
```

File: spec/xtrabackuprb_spec.rb [hunk 15]
```diff
@@ -0,0 +1,88 @@
+require 'spec_helper'
+
+describe 'osl-mysql::xtrabackuprb' do
+  ALLPLATFORMS.each do |pltfrm|
+    context "on #{pltfrm[:platform]} #{pltfrm[:version]}" do
+      cached(:chef_run) do
+        ChefSpec::SoloRunner.new(pltfrm).converge(described_recipe)
+      end
+      it do
```
Comment (ramereth) [inline spec/xtrabackuprb_spec.rb]: You need to add some stubs here so that rspec runs. This also needs to happen on the backup recipe and might make sense to have this as a shared_context so you don't duplicate code.

**Comment by ramereth** [comment]
@JerryPeng0112 can you also update the test provisioning environment (i.e. ``test/integration/provisioning/master_slave.rb``). Please do the following to that file:
- Update to CentOS 7.5
- Remove the box_version line
- Update Chef version to 13.8.5
- Test by running ``rake master_slave``

**Comment by ramereth** [comment]
@JerryPeng0112 you need to do the following to get this to work on Chef 13:
``` diff
diff --git a/Rakefile b/Rakefile
index 82d126b..68a2334 100644
--- a/Rakefile
+++ b/Rakefile
@@ -12,7 +12,7 @@ snakeoil_file_path = 'test/integration/data_bags/certificates/snakeoil.json'
 encrypted_data_bag_secret_path = 'test/integration/encrypted_data_bag_secret'
 current_dir = File.dirname(__FILE__)
 client_cfg = "#{current_dir}/test/chef-config"
-client_options = '--force-formatter -z ' \
+client_options = '--force-formatter -z --local-mode --listen ' \
     "--config #{client_cfg}/knife.rb"
 
 ##
```

@osuosl-cookbooks/chefs take note on other Chef 13 PR's ^^^

**Comment by ramereth** [comment]
@JerryPeng0112 any update on this?

**Comment by ramereth** [comment]
@doublej472 looks like the ``mon-centos-7`` suite is failing otherwise everything looks good to me.

**Comment by doublej472** [comment]
I've gotten the failing test to work somewhat (the monitor user was missing privileges), but line `1859` in `/etc/munin/plugins/mysql_bin_relay_log` is deleting the `bin_relay_log` graph because `relay_log_space` isn't defined:
```perl
delete $graphs{bin_relay_log} if not defined $data->{relay_log_space};
```
Which then dies later on line `1805` because the graph is missing
```perl
die 'Unknown graph ' . ($graph_name ? $graph_name : '')
    unless $graphs{$graph_name};
```

With the `delete` line removed I can get partially usable output (basically anything that is in the binary log and NOT the relay log):
```
[root@moncentos7-doublej472-crimsonworkstationcasso-2a0y7vr ~]# /usr/sbin/munin-run mysql_bin_relay_log
Binlog_cache_disk_use.value 0
Binlog_cache_use.value 0
Binlog_stmt_cache_disk_use.value 0
Binlog_stmt_cache_use.value 11
ma_binlog_size.value 1869
relay_log_space.value U
```

The relay log only exists for [slave replicated mysql instances](https://dev.mysql.com/doc/refman/5.7/en/slave-logs-relaylog.html), so this test might not be appropriate for single node mysql instances, such as the test kitchen instance. It would make more sense to test this in provisioning, but I don't know how much work I should put into this.

@ramereth what do you think?

**Comment by ramereth** [comment]
@doublej472 let's leave it out for now then and just comment it out with a reason.

**Comment by doublej472** [comment]
@ramereth I don't think there is a real way to add intergration tests that we aren't already testing in `osl-mysql::server`, since it's is literally the same thing, only with configuration values changed so it acts as a slave mysql instance, which can't work in a single-node test.

I can check if proper values are set in the recipe, but that is a unit test, not an integration test, which goes in chefspec. Another option is to create a provisioning setup, but that is already done in the Rakefile.

**Comment by doublej472** [comment]
!bump major ~

**Comment by osuosl-manatee** [comment]
Jenkins has merged this PR into `master` and has automatically performed a major-level version bump to v2.0.0.  Have a nice day!
